core/interop/src/loaders/embedded.rs

```diff
@@ -0,0 +1,141 @@
+//! Embedded module loader. Creates a `ModuleLoader` instance that contains
+//! files embedded in the binary at build time.
+
+use std::cell::RefCell;
+use std::collections::HashMap;
+use std::path::Path;
+
+use boa_engine::module::{ModuleLoader, Referrer};
+use boa_engine::{Context, JsError, JsNativeError, JsResult, JsString, Module, Source};
+
+/// Create a module loader that embeds files from the filesystem at build
+/// time. This is useful for bundling assets with the binary.
+///
+/// By default, will error if the total file size exceeds 1MB. This can be
+/// changed by specifying the `max_size` parameter.
+///
+/// The embedded module will only contain files that have the `.js`, `.mjs`,
+/// or `.cjs` extension.
+#[macro_export]
+macro_rules! embed_module {
+    ($path: literal, max_size = $max_size: literal) => {
+        $crate::loaders::embedded::EmbeddedModuleLoader::from_iter(
+            $crate::boa_macros::embed_module_inner!($path, $max_size),
+        )
+    };
+    ($path: literal) => {
+        embed_module!($path, max_size = 1_048_576)
+    };
+}
+
+#[derive(Debug, Clone)]
+enum EmbeddedModuleEntry {
+    Source(JsString, &'static [u8]),
+    Module(Module),
+    Error(JsError),
```

Comment:
Can you explain why do we need an `Error` variant on the entry?
I would expect that `cache` would throw if parsing fails, instead of caching the fail.

Reply:
Optimize the case where the same module is re-imported? There's no need to reparse it.

Reply:
Yeah, but it's not common to cache the parse error because in most cases execution will bail on the first error.

Reply:
Sure, removed.